dll/win32/lpk/lpk.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * PROJECT:     ReactOS LPK
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     Language Pack DLL.
+ * PROGRAMMERS: Magnus Olsen  (greatlrd)
+ *              Baruch Rutman (peterooch at gmail dot com)
+ */
+
+#include "ros_lpk.h"
+
+LPK_LPEDITCONTROL_LIST LpkEditControl = {EditCreate,       EditIchToXY,  EditMouseToIch, EditCchInWidth,
+                                         EditGetLineWidth, EditDrawText, EditHScroll,    EditMoveSelection,
+                                         EditVerifyText,   EditNextWord, EditSetMenu,    EditProcessMenu,
+                                         EditCreateCaret, EditAdjustCaret};
+
+BOOL
+WINAPI
+DllMain(
+    HANDLE  hDll,
+    DWORD   dwReason,
+    LPVOID  lpReserved)
+{
+
+    return LpkDllInitialize(hDll,dwReason,lpReserved);
+}
+
+BOOL
+WINAPI
+LpkDllInitialize(
+    HANDLE  hDll,
+    DWORD   dwReason,
+    LPVOID  lpReserved)
+{
+    switch(dwReason)
+    {
+        case DLL_PROCESS_ATTACH:
+            DisableThreadLibraryCalls(hDll);
+            /* Tell usp10 it is activated usp10 */
+            //LpkPresent();
+            break;
+
+        default:
+            break;
+    }
+
+    return TRUE;
+}
+
+
+/*
+ * @implemented 
+ */
+BOOL
+WINAPI
+LpkExtTextOut(
+    HDC hdc,
+    int x,
+    int y,
+    UINT fuOptions,
+    const RECT *lprc,
+    LPCWSTR lpString,
+    UINT uCount,
+    const INT *lpDx,
+    INT unknown)
+{
+    WORD *glyphs = NULL;
+    INT cGlyphs;
+
+    UNREFERENCED_PARAMETER(unknown);
+
+    /* Check if the string requires complex script processing and not a "glyph indices" array */
+    if (ScriptIsComplex(lpString, uCount, SIC_COMPLEX) == S_OK && !(fuOptions & ETO_GLYPH_INDEX))
+    {
+        BIDI_Reorder(hdc, lpString, uCount, GCP_REORDER, WINE_GCPW_FORCE_LTR,
+                     NULL, uCount, NULL, &glyphs, &cGlyphs);
+
+        fuOptions |= ETO_GLYPH_INDEX;
+
+        if (uCount > cGlyphs)
+        cGlyphs = uCount;
+
+        return ExtTextOutW(hdc, x, y, fuOptions, lprc, (LPWSTR)glyphs, cGlyphs, lpDx);
+    }
+
+    if (!(fuOptions & ETO_IGNORELANGUAGE))
+        fuOptions |= ETO_IGNORELANGUAGE;
+
+    return ExtTextOutW(hdc, x, y, fuOptions, lprc, lpString, uCount, lpDx);
+}
+
+/*
+ * @implemented
+ */
+DWORD
+WINAPI
+LpkGetCharacterPlacement(
+    HDC hdc,
+    LPCWSTR lpString,
+    INT uCount,
+    INT nMaxExtent,
+    GCP_RESULTSW *lpResults,
+    DWORD dwFlags,
+    DWORD dwUnused)
+{
+    LPWSTR lpOutString;
+    WORD *lpGlyphs = NULL;
+    UINT *lpOrder = NULL;
+    SIZE size;
+    DWORD ret = 0;
+    UINT nSet, i;
+    INT cGlyphs;
+
+    UNREFERENCED_PARAMETER(dwUnused);
+
+    /* Sanity check (most likely a direct call) */
+    if (!(dwFlags & GCP_REORDER))
+       return GetCharacterPlacementW(hdc, lpString, uCount, nMaxExtent, lpResults, dwFlags);
+
+    lpOutString = HeapAlloc(GetProcessHeap(), 0, uCount * sizeof(WCHAR));
```

Comment:
Need to check for allocation failure (and return some adequate value).

Reply:
I don't know, but the usual programming pattern is that whenever you fail allocating your buffer when you need it afterwards, you should fail properly (or at least take another code path).

Here you are making an hypothesis (?) about how the BIDI_reorder function works (it handles NULL pointers certainly to handle cases when the user explicitly calls it with NULL, because it may not need the info in return, etc...) If however you really need that info then you pass some non-NULL pointer. Here it seems, at least, that you really want to receive some info in lpOutString, so, you would prefer not to fail the allocation for it!

If on the contrary you actually don't care about what could be returned in lpOutString *and* you know that BIDI_reorder returns adequate results even if its lpOutString parameter is NULL, then *maybe* you just have to pass a NULL pointer and that's it (OK: I haven't checked its code so what I've just told in this paragraph is an hypothesis).

Reply:
`BIDI_Reorder(hdc, lpString, uCount, dwFlags, WINE_GCPW_FORCE_LTR, lpOutString ? lpOutString : NULL,` is this enough of a check?

Reply:
Hmmm....
... Actually it looks like you should directly pass lpResults->lpOutString and lpResults->lpOrder pointers directly to BIDI_reorder, instead of doing some temporary allocation... ?

> GetCharacterPlacement checks the lpOrder, lpDX, lpCaretPos, lpOutString, and lpGlyphs members of the GCP_RESULTS structure and fills the corresponding arrays if these members are not set to NULL. If GetCharacterPlacement cannot fill an array, it sets the corresponding member to NULL. To ensure retrieval of valid information, the application is responsible for setting the member to a valid address before calling the function and for checking the value of the member after the call. If the GCP_JUSTIFY or GCP_USEKERNING values are specified, the lpDX and/or lpCaretPos members must have valid addresses.
[GetCharacterPlacement function](https://msdn.microsoft.com/en-us/library/windows/desktop/dd144860(v=vs.85).aspx)